File: Python/test_pmangplus.py
# -*- coding: utf-8 -*-

import unittest
import json
import requests
import time
import hashlib
import random
import socket

# application identifier
APP_ID = 'pmang app id'
APP_KEY = 'pmang app key'
APP_SECRET = 'pmang app secret'
LOCAL_CD = 'KOR'
DEVICE_CD = 'ANDROID'
#DEVICE_CD = 'IPHONE'

# user identifier
ID = 'pmang id'
PW = 'pmang pw'

# payment identifier
DEVELOPER_PAYLOAD = 'payment developer payload'
PURCHASE_TOKEN = 'payment purchase token'

# QA/TQ/Live URL
URL = 'http://qa.pmangplus.com'
#URL = 'http://tq.pmangplus.com'
#URL = 'http://pmangplus.com'


class TestClasses(unittest.TestCase):

	_access_token = None
	_member_id = None
	_nickname = None


	def setUp(self):
		def _params(email, passwd):
			udid = 'TestUDID_' + socket.gethostbyname(socket.gethostname())
			return {
				'local_cd': LOCAL_CD,
				'app_id': APP_ID,
				'app_key': APP_KEY,
				'app_secret': APP_SECRET,
				'udid': udid,
				'email': email,
				'passwd': passwd,
				'device_cd': DEVICE_CD
			}

		params = _params(ID, PW)
		res = self._post('/accounts/2/login', params)

		assert res['value']['access_token']
		assert res['value']['member']['member_id']
		assert res['value']['member']['nickname']

		self._access_token = res['value']['access_token']
		self._member_id = str(res['value']['member']['member_id'])
		self._nickname = res['value']['member']['nickname']

	@unittest.skip('')
	def test_self(self):
		def _params():
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token
			}
		params = _params()
		res = self._get('/members/@self', params)

		self.assertTrue(res['value'].has_key('crt_dt'))
		self.assertTrue(res['value'].has_key('upd_dt'))
		self.assertEqual(str(res['value']['member_id']), self._member_id)
		self.assertEqual(res['value']['nickname'], self._nickname)
		self.assertTrue(res['value'].has_key('profile_img_url'))
		self.assertTrue(res['value'].has_key('feeling'))
		self.assertTrue(res['value'].has_key('adult_auth_yn'))
		self.assertTrue(res['value'].has_key('adult_auth_dt'))
		self.assertTrue(res['value'].has_key('recent_login_dt'))
		self.assertTrue(res['value'].has_key('recent_app_id'))
		self.assertTrue(res['value'].has_key('email'))
		self.assertTrue(res['value'].has_key('anonymous_yn'))
		self.assertTrue(res['value'].has_key('reg_path'))
		self.assertTrue(res['value'].has_key('recent_app_title'))
		self.assertTrue(res['value'].has_key('last_msg_dt'))
		self.assertTrue(res['value'].has_key('new_msg_yn'))
		self.assertTrue(res['value'].has_key('friend_accept_cd'))
		self.assertTrue(res['value'].has_key('achieve_detail_info_list'))
		self.assertTrue(res['value'].has_key('member_achievement_summary'))
		self.assertTrue(res['value'].has_key('summary'))
		self.assertTrue(res['value'].has_key('options'))
		self.assertTrue(res['value'].has_key('contact_require_phone_auth'))
		self.assertTrue(res['value'].has_key('contact_require_import'))
		self.assertTrue(res['value'].has_key('contact_friend_mapping'))
		self.assertTrue(res['value'].has_key('contact_require_phone_auth_confirm'))
		self.assertTrue(res['value'].has_key('contact'))
		self.assertTrue(res['value'].has_key('pmang_usn'))
		self.assertTrue(res['value'].has_key('ci'))
		self.assertTrue(res['value'].has_key('sanction'))
		self.assertTrue(res['value'].has_key('profile_img_url_raw'))


	@unittest.skip('')
	def test_payment(self):
		def _params(developer_payload, purchase_token):
			return {
				'local_cd': LOCAL_CD,
				'developerPayload': developer_payload,
				'purchaseToken': purchase_token
			}
		params = _params(DEVELOPER_PAYLOAD, PURCHASE_TOKEN)
		res = self._post('/market/v3/effectuate', params)

		# if success
		#self.assertEqual(res['value']['result'], 'SUCCESS')
		#self.assertTrue(res['value'].has_key('store_product_id'))
		#self.assertTrue(res['value'].has_key('pay_id'))

		# if no success(FAILURE / DUPLICATED / NOTEXISTS)
		self.assertTrue(res['value'].has_key('result'))


	@unittest.skip('')
	def test_gcm_publish(self):
		def _params(member_ids):
			return {
				'local_cd': LOCAL_CD,
				'member_ids': ','.join(member_ids),
				'message': """{
					'content':'내용',
					'title':'제목',
					'type':'ad',
					'url':'http://ppl.cm/abcdef'
				}"""
			}
		member_ids = (str(self._member_id), str(self._member_id), str(self._member_id))
		params = _params(member_ids)
		res = self._post('/gcm/' + APP_ID + '/publish', params)

		self.assertTrue(res.has_key('value'))


	@unittest.skip('')
	def test_gcm_get_allow(self):
		def _params():
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token
			}
		params = _params()
		res = self._get('/gcm/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))


	@unittest.skip('')
	def test_gcm_set_allow(self):
		def _get_params():
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token
			}

		def _set_params(allow):
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token,
				'allow': allow
			}

		# get allow
		params = _get_params()
		res = self._get('/gcm/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))
		push_allow = res['value']

		# toggle allow
		params = _set_params(not push_allow)
		res = self._post('/gcm/' + APP_ID + '/setAllow', params)

		self.assertTrue(res.has_key('value'))
		self.assertEqual(not push_allow, res['value'])

		# confirm allow
		params = _get_params()
		res = self._get('/gcm/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))
		self.assertEqual(not push_allow, res['value'])


	def test_apns_publish(self):
		def _params(member_ids):
			return {
				'local_cd': LOCAL_CD,
				'member_ids': ','.join(member_ids),
				'message': '메세지'
			}
		member_ids = (str(self._member_id), str(self._member_id), str(self._member_id))
		params = _params(member_ids)
		res = self._post('/apns/' + APP_ID + '/publish', params)

		self.assertTrue(res.has_key('value'))


	def test_apns_get_allow(self):
		def _params():
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token
			}
		params = _params()
		res = self._get('/apns/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))


	def test_apns_set_allow(self):
		def _get_params():
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token
			}

		def _set_params(allow):
			return {
				'local_cd': LOCAL_CD,
				'access_token': self._access_token,
				'allow': allow
			}

		# get allow
		params = _get_params()
		res = self._get('/apns/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))
		push_allow = res['value']

		# toggle allow
		params = _set_params(not push_allow)
		res = self._post('/apns/' + APP_ID + '/setAllow', params)

		self.assertTrue(res.has_key('value'))
		self.assertEqual(not push_allow, res['value'])

		# confirm allow
		params = _get_params()
		res = self._get('/apns/' + APP_ID + '/getAllow', params)
		
		self.assertTrue(res.has_key('value'))
		self.assertEqual(not push_allow, res['value'])



	def _make_header(self):
		mills = int(time.time())

		h = hashlib.sha1()
		h.update(str(mills) + APP_SECRET)
		digest = h.hexdigest()

		return {
			'ver': '4',
			'ts': str(mills),
			'fp': digest,
			'User-Agent': 'PmangPlus for Server',
			'content-type': 'application/json;charset=UTF-8'
		}

	def _post(self, prefix, params):
		url = URL + prefix
		headers = self._make_header()
		response = requests.post(url, params=params, headers=headers)
		res = json.loads(response.content)
		self.assertEqual(res['result_code'], '000')
		self.assertEqual(res['result_msg'], 'API_OK')
		return res

	def _get(self, prefix, params):
		url = URL + prefix
		headers = self._make_header()
		r = requests.get(url, params=params, headers=headers)
		res = json.loads(r.content)
		self.assertEqual(res['result_code'], '000')
		self.assertEqual(res['result_msg'], 'API_OK')
		return res

File: Node.js/test/test-pmangplus.js
'use strict';

var request = require('request');
var assert = require('assert');
var async = require('async');
var sha1 = require('sha1');

// application identifier
var APP_ID = 'pmang app id';
var APP_KEY = 'pmang app key';
var APP_SECRET = 'pmang app secret';
var LOCAL_CD = 'KOR';
var DEVICE_CD = 'ANDROID';
//var DEVICE_CD = 'IPHONE'

// user identifier
var ID = 'pmang id';
var PW = 'pmang pw';

// payment identifier
var DEVELOPER_PAYLOAD = 'payment developer payload';
var PURCHASE_TOKEN = 'payment purchase token';

// QA/TQ/Live URL
var URL = 'http://qa.pmangplus.com';
//var URL = 'http://tq.pmangplus.com';
//var URL = 'http://pmangplus.com';


describe('P+ TestCase', function() {

	var access_token = '';
	var member_id = '';
	var nickname = '';
	var push_allow = false;

	before(function(done) {
		// login
		async.series([
			function (callback) {
				var options = post('/accounts/2/login');
				options.form = {
						'local_cd': LOCAL_CD,
						'app_id': APP_ID,
						'app_key': APP_KEY,
						'app_secret': APP_SECRET,
						'udid': 'TestUDID_Node',
						'email': ID,
						'passwd': PW,
						'device_cd': DEVICE_CD
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);
					assert.ok(res.value.access_token);
					assert.ok(res.value.member.member_id);
					assert.ok(res.value.member.nickname);

					access_token = res.value.access_token;
					member_id = res.value.member.member_id;
					nickname = res.value.member.nickname;

					done();
				});
			}
		], done);
	});


	it('test members @self', function(done) {
		async.series([
			function (callback) {
				var options = post('/members/@self');
				options.form = {
						'local_cd': LOCAL_CD,
						'access_token': access_token
				};
		
				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.value.has('crt_dt') && res.value.crt_dt);
					assert.ok(res.value.has('upd_dt') && res.value.upd_dt);
					assert.ok(res.value.has('status_cd') && res.value.status_cd);
					assert.deepStrictEqual(res.value.member_id, member_id);
					assert.deepStrictEqual(res.value.nickname, nickname);
					assert.ok(res.value.has('profile_img_url'));
					assert.ok(res.value.has('feeling'));
					assert.ok(res.value.has('adult_auth_yn') && res.value.adult_auth_yn);
					assert.ok(res.value.has('adult_auth_dt'));
					assert.ok(res.value.has('recent_login_dt'));
					assert.ok(res.value.has('recent_app_id'));
					assert.ok(res.value.has('email'));
					assert.ok(res.value.has('anonymous_yn'));
					assert.ok(res.value.has('reg_path'));
					assert.ok(res.value.has('recent_app_title'));
					assert.ok(res.value.has('last_msg_dt'));
					assert.ok(res.value.has('new_msg_yn'));
					assert.ok(res.value.has('friend_accept_cd'));
					assert.ok(res.value.has('achieve_detail_info_list'));
					assert.ok(res.value.has('member_achievement_summary'));
					assert.ok(res.value.has('summary'));
					assert.ok(res.value.has('options'));
					assert.ok(res.value.has('contact_require_phone_auth'));
					assert.ok(res.value.has('contact_require_import'));
					assert.ok(res.value.has('contact_friend_mapping'));
					assert.ok(res.value.has('contact_require_phone_auth_confirm'));
					assert.ok(res.value.has('contact'));
					assert.ok(res.value.has('pmang_usn'));
					assert.ok(res.value.has('ci'));
					assert.ok(res.value.has('sanction'));
					assert.ok(res.value.has('profile_img_url_raw'));

					done();
				});
			}
		], done);
	});


	it('test market effectuate', function(done) {
		async.series([
			function (callback) {
				var options = post('/market/v3/effectuate');
				options.form = {
					'local_cd': LOCAL_CD,
					'developerPayload': DEVELOPER_PAYLOAD,
					'purchaseToken': PURCHASE_TOKEN
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					// if success
					//assert.deepStrictEqual(res.value.result, 'SUCCESS');
					//assert.ok(res.value.has('store_product_id') && res.value.store_product_id);
					//assert.ok(res.value.has('pay_id') && res.value.pay_id);

					// if no success(FAILURE / DUPLICATED / NOTEXISTS)
					assert.ok(res.value.has('result') && res.value.result);

					done();
				});
			}
		], done);
	});


	it('test gcm publish', function(done) {
		async.series([
			function (callback) {
				// member_id,member_id,member_id
				var member_ids = [member_id, member_id, member_id];
				var options = post('/gcm/' + APP_ID + '/publish');
				options.form = {
					'local_cd': LOCAL_CD,
					'member_ids': member_ids.join(),
					'message': '{"content":"내용", "title":"제목", "type":"ad", "url":"http://ppl.cm/abcdef"}'
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.value);

					done();
				});
			}
		], done);
	});


	it('test gcm get allow', function(done) {
		async.series([
			function (callback) {
				var options = post('/gcm/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));

					done();
				});
			}
		], done);
	});


	it('test gcm set allow', function(done) {
		async.series([
			// get allow
			function (callback) {
				var options = post('/gcm/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					push_allow = res.value;

					done();
				});
			},
			// toggle allow
			function (callback) {
				var options = post('/gcm/' + APP_ID + '/setAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token,
					'allow': !push_allow
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					assert.equal(!push_allow, res.value);

					done();
				});
			},
			// confirm allow
			function (callback) {
				var options = post('/gcm/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					assert.equal(!push_allow, res.value);

					done();
				});
			}
		], done);
	});


	it('test apns publish', function(done) {
		async.series([
			function (callback) {
				// member_id,member_id,member_id
				var member_ids = [member_id, member_id, member_id];
				var options = post('/apns/' + APP_ID + '/publish');
				options.form = {
					'local_cd': LOCAL_CD,
					'member_ids': member_ids.join(),
					'message': 'my message'
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.value);

					done();
				});
			}
		], done);
	});


	it('test apns get allow', function(done) {
		async.series([
			function (callback) {
				var options = post('/apns/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));

					done();
				});
			}
		], done);
	});


	it('test apns set allow', function(done) {
		async.series([
			// get allow
			function (callback) {
				var options = post('/apns/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					push_allow = res.value;

					done();
				});
			},
			// toggle allow
			function (callback) {
				var options = post('/apns/' + APP_ID + '/setAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token,
					'allow': !push_allow
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					assert.equal(!push_allow, res.value);

					done();
				});
			},
			// confirm allow
			function (callback) {
				var options = post('/apns/' + APP_ID + '/getAllow');
				options.form = {
					'local_cd': LOCAL_CD,
					'access_token': access_token
				};

				request(options, function (error, response, body) {
					var res = get_response(error, response, body);

					assert.ok(res.hasOwnProperty('value'));
					assert.equal(!push_allow, res.value);

					done();
				});
			}
		], done);
	});
	
});


function make_header() {
	var now = Date.now();
	return {
		'ver': 4,
		'ts': now,
		'fp': sha1(now + APP_SECRET),
		'User-Agent': 'PmangPlus for Server',
		'Content-Type': 'application/json;charset=UTF-8'};
}

function post(prefix) {
	return {
		url: URL + prefix,
		method: 'POST',
		headers: make_header()
	}
}

function get(prefix) {
	return {
		url: URL + prefix,
		method: 'GET',
		headers: make_header()
	}
}

function get_response(error, response, body) {
	if (error) assert.fail(error);
	if (response.statusCode != 200) assert.fail(response);

	var res = JSON.parse(body);
	if (typeof res.value == 'object') {
		res.value.has = function (name) {
			return this.hasOwnProperty(name);
		};
	}

	assert.deepStrictEqual(res.result_code, '000');
	assert.deepStrictEqual(res.result_msg, 'API_OK');

	return res;
}
